### PR TITLE
Focus input when InputControl spinner arrows are pressed

### LINKED ISSUE #29271: Components: NumberControl up/down buttons not working correctly
## Description

When selecting the columns block in Gutenberg (in the Firefox or Safari), the arrow buttons in the block inspector to add or remove additional columns is broken

## Step-by-step reproduction instructions

1. Navigate to the post editor or site editor
2. Add a columns block
3. Open the block inspector
4. Verify that the arrow buttons does not adjust the number of columns as expected

## Expected behaviour

- Clicking on the "up" arrow will add an additional column to the columns block
- Clicking down on the "down" arrow will remove a column from the columns block

## Actual behaviour

The arrow buttons are unresponsive or don't function at as expected

## WordPress information
- WordPress version: 
- Gutenberg version: `master`
- Are all plugins except Gutenberg deactivated? Yes
- Are you using a default theme (e.g. Twenty Twenty-One)? TT1 blocks

## Device information
- Device: Desktop
- Operating system: Mac Big Sur
- Browser: Firefox latest

### PR DESCRIPTION
**Pull request by stokesman**
Fix #29271

This has been done before 4e7c772014478cf0aec58cb1da6cc42fe17710f8. It was removed in bf9f9e572138ab28814c1b113a6d9b5fbbfaea8c. This brings it back (and makes sure the drag-to-adjust feature still works).

UPDATE: The drag-to-adjust feature is not working in Safari (versions <= 12) but is unrelated to this PR. Also with regards to Safari support:
>According to this doc, [we support the last two versions ](https://developer.wordpress.org/block-editor/handbook/faq/#what-browsers-does-gutenberg-support)

Thanks @jeyip! 

## How has this been tested?
In Firefox, Chrome, and Safari, in the post editor, with a range control, clicking the up/down arrows of the number input before having focused the input. Also dragging the input field to adjust to ensure that still works.

## Types of changes
Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by jeyip** [comment]
## Testing
### Behaviors
- [x] Drag-to-adjust functions correctly
- [x] Up and down arrow keys increment and decrement correctly
- [x] Spinner arrows increment and decrement correctly
- [x] Manual number adjustment with the keyboard functions correctly

###  
- [x] Firefox
- [x] Edge
- [x] Chrome
- [ ] ~Safari~

**Note:**
As mentioned by @stokesman, the behavior in Safari remains broken. The issue, however, appears to be distinct from the problem we're solving here.

**Comment by stokesman** [comment]
Thanks for testing @jeyip!

I just realized that my version of Safari is 12.1.2 and https://caniuse.com/mdn-api_globaleventhandlers_onpointerdown says the pointer events aren't supported there. The newer version of react-use-gesture is all pointer events so apparently only supports Safari 13 and up. I don't know if Gutenberg has any policy for Safari support. If not, looks like we can let it slide.

File: packages/components/src/input-control/input-field.js [hunk 1]
```diff
@@ -184,6 +184,18 @@ function InputField(
 	);
 
 	const dragProps = isDragEnabled ? dragGestureProps() : {};
+	/*
+	 * Works around the odd UA (e.g. Firefox) that does not focus inputs of
+	 * type=number when their spinner arrows are pressed.
+	 */
+	let handleOnMouseDown;
```
Comment (jeyip) [inline packages/components/src/input-control/input-field.js:191]: Could we initialize this as a `noop`?

`let handleOnMousedown = noop;`

**Review by jeyip — COMMENTED** [review]
Would it be possible to add unit tests for this?

**Review by jeyip — COMMENTED** [review]
@ItsJonQ This looks reasonable to me, but it'd be awesome if you could double check this and make sure we're not missing anything obvious 🙂

**Review by jeyip — COMMENTED** [review]
+1 from me whenever we resolve all the comments 🚀

**Comment by jeyip** [comment]
> I just realized that my version of Safari is 12.1.2 and https://caniuse.com/mdn-api_globaleventhandlers_onpointerdown says the pointer events aren't supported there. The newer version of react-use-gesture is all pointer events so apparently only supports Safari 13 and up. I don't know if Gutenberg has any policy for Safari support.

That's a great question. According to this doc, [we support the last two versions ](https://developer.wordpress.org/block-editor/handbook/faq/#what-browsers-does-gutenberg-support)

**Comment by stokesman** [comment]
>According to this doc, we support the last two versions

I managed to test this in Safari 13.1 by way of Browserstack and gutenberg.run 😎 . The dragging feature worked there. Looks like the latest version is 14, so if my math is correct we can not sweat the lack of Safari 12 support (not that the lack of this feature actually busted anything).

File: packages/components/src/input-control/input-field.js (hunk 1 above)
Comment (stokesman) [reply packages/components/src/input-control/input-field.js:191]: Certainly, though I'm unaware of the benefit and would be curious to learn. I'm under the impression that when this remains `undefined` React won't bother to add any listener when it's passed as the `onMouseDown` prop. If that's correct, I still don't feel strongly about it.

**Comment by stokesman** [comment]
> Would it be possible to add unit tests for this?

Good question. I think it might not be since it's pretty specific to Firefox. The relief should be, I think, that before too long all `InputControl` based components will be phased out by something from g2 which will not be subject to this little quirk.

**Comment by jeyip** [comment]
> I think it might not be since it's pretty specific to Firefox.

What if we didn't make the test firefox specific, simulated a mousedown event on the input field, spied on the `event.target.focus()` method and ensured that focus was invoked?

With that being said, I think you bring up a good point with the introduction of G2 components. I think there's value in the unit test preventing this bug from being reintroduced, but, depending on the time horizon, I'm comfortable skipping it and waiting for these components to be phased out.

**Comment by stokesman** [comment]
Thanks for the nudge @jeyip. I'd totally thought a test like that would never fail but it does indeed without the change in this PR. For minutiae, I couldn't see a way to spy on `event.target.focus`. I did find it worked to assert `event.target === activeElement` but that felt unnecessarily explicit when the simper test I've pushed here does the trick.

**Review by ItsJonQ — APPROVED** [review]
🚀 from me! Thank you @stokesman !
I tested it in Chrome, Safari, and FF. The focus works for me.
However, there seems to be something odd with Safari + spinner for me.
I think it's the [`react-use-gesture`](https://github.com/pmndrs/react-use-gesture) library 🤔 .

If it helps, I believe these interaction issues will be resolved when we update these inputs with ones from the [new Component System](https://github.com/WordPress/gutenberg/issues/28399).

@sarayourfriend and I are getting closer to adding these input components.

As it relates to this issue, the new inputs use custom "spinners", which bypasses some of the different browser arrow interaction oddities. It also supports features like `shift` step jumping and incrementing/decrementing with "velocity" (going up/down the longer you hold it).

Demo:
https://g2-components.xyz/iframe.html?id=components-textinput--number&viewMode=story

**Review by jeyip — APPROVED** [review]

File: packages/components/src/input-control/input-field.js (hunk 1 above)
Comment (jeyip) [reply packages/components/src/input-control/input-field.js:191]: Oh it was just to align with the conventions from the components input props. When not defined, the default params are `noop` for the event handlers

Screen Shot 2021-02-25 at 11 06 18 AM

Comment (stokesman) [reply packages/components/src/input-control/input-field.js:191]: Right, it's not quite the same thing but it was thanks to looking trying to understand this, that I noticed that my changes would have prevented the propagation out to a `onMouseDown` from props. So I pushed a451b11813a3d1dffb61562cdfd8ad7f60527d51. Instead, adding `onMouseDown = noop` along with those others I just used optional chaining for the propagation call which negates the need to initialize to `noop`.

**Comment by jeyip** [comment]
Everything looks good to me! Feel free to merge when ready @stokesman 🚀 

Thanks again for the help with this 🙏

**Pull request merged**